File: ep09-transform-fk-wrapup/python/tf.py
"""EP.9 공구함 모듈 tf.py — 2D 강체변환 · 동차변환 체인 · 다링크 FK · DH · 스윕 · 각도 도구 · SAT.

강의 S33·S34가 약속한 함수군을 한 파일에 둔다. 각도 인자는 모두 도(degree)이며
함수 안에서만 라디안으로 바꾼다. 외부 라이브러리 없이 math만 쓴다.
"""
import math

# ---------- 각도 도구 ----------

def wrap(deg):
    """각도를 [-180, 180) 범위로 접는다. wrap(-325) = 35."""
    return (deg + 180.0) % 360.0 - 180.0


def circular_mean(degs):
    """원형 평균 = atan2(sum sin, sum cos). 10·350·20도 → 약 6.7도(단순 평균 126.6667은 오답)."""
    s = sum(math.sin(math.radians(d)) for d in degs)
    c = sum(math.cos(math.radians(d)) for d in degs)
    return math.degrees(math.atan2(s, c))


# ---------- 2D 강체변환 ----------

def rot2(deg):
    c, s = math.cos(math.radians(deg)), math.sin(math.radians(deg))
    return [[c, -s], [s, c]]


def rigid2d(points, deg, t):
    """p' = R p + t (회전 후 이동). 리스트 컴프리헨션 한 줄."""
    (a, b), (c, d) = rot2(deg)
    return [(a * x + b * y + t[0], c * x + d * y + t[1]) for x, y in points]


def inv2d(deg, t):
    """역변환의 (각도, 이동항). 이동항은 -R^T t 이지 단순 -t 가 아니다."""
    (a, b), (c, d) = rot2(deg)
    return -deg, (-(a * t[0] + c * t[1]), -(b * t[0] + d * t[1]))


def recover_angle(p_orig, p_new, t):
    """변환된 한 점에서 회전각 복원: (atan2 각, 내적·acos 각). 변환의 역이지 역기구학이 아니다."""
    vx, vy = p_new[0] - t[0], p_new[1] - t[1]
    by_atan2 = math.degrees(math.atan2(vy, vx)) - math.degrees(math.atan2(p_orig[1], p_orig[0]))
    dot = p_orig[0] * vx + p_orig[1] * vy
    norm = math.hypot(*p_orig) * math.hypot(vx, vy)
    by_acos = math.degrees(math.acos(max(-1.0, min(1.0, dot / norm))))
    return by_atan2, by_acos


# ---------- 4x4 동차변환 ----------

def rotz(deg):
    c, s = math.cos(math.radians(deg)), math.sin(math.radians(deg))
    return [[c, -s, 0.0], [s, c, 0.0], [0.0, 0.0, 1.0]]


def rotx(deg):
    c, s = math.cos(math.radians(deg)), math.sin(math.radians(deg))
    return [[1.0, 0.0, 0.0], [0.0, c, -s], [0.0, s, c]]


def homog(R, t):
    """[R t; 0 1] 4x4 조립."""
    return [[R[0][0], R[0][1], R[0][2], t[0]],
            [R[1][0], R[1][1], R[1][2], t[1]],
            [R[2][0], R[2][1], R[2][2], t[2]],
            [0.0, 0.0, 0.0, 1.0]]


def mul(A, B):
    return [[sum(A[i][k] * B[k][j] for k in range(4)) for j in range(4)] for i in range(4)]


def apply(T, p):
    return tuple(T[i][0] * p[0] + T[i][1] * p[1] + T[i][2] * p[2] + T[i][3] for i in range(3))


def inv4(T):
    """[R^T  -R^T t; 0 1]."""
    Rt = [[T[j][i] for j in range(3)] for i in range(3)]
    t = [-(Rt[i][0] * T[0][3] + Rt[i][1] * T[1][3] + Rt[i][2] * T[2][3]) for i in range(3)]
    return homog(Rt, t)


# ---------- 다링크 순기구학 ----------

def fk(links, angles_deg):
    """상대 관절각을 누적해 원점부터 끝점까지 관절 좌표 목록(튜플)을 반환."""
    x = y = cum = 0.0
    pts = [(0.0, 0.0)]
    for l, a in zip(links, angles_deg):
        cum += a
        x += l * math.cos(math.radians(cum))
        y += l * math.sin(math.radians(cum))
        pts.append((x, y))
    return pts


def sweep(links, theta1_deg, start, stop, step):
    """theta1 고정, theta2 를 start..stop(양 끝 포함) 로 훑는 FK 반복."""
    return [(a, fk(links, (theta1_deg, a))[-1]) for a in range(start, stop + 1, step)]


# ---------- DH ----------

def dh(a, alpha_deg, d, theta_deg):
    """표준 DH 행렬 A_i = Rz(theta) Tz(d) Tx(a) Rx(alpha)."""
    ca, sa = math.cos(math.radians(alpha_deg)), math.sin(math.radians(alpha_deg))
    ct, st = math.cos(math.radians(theta_deg)), math.sin(math.radians(theta_deg))
    return [[ct, -st * ca, st * sa, a * ct],
            [st, ct * ca, -ct * sa, a * st],
            [0.0, sa, ca, d],
            [0.0, 0.0, 0.0, 1.0]]


def dh_chain(rows):
    """[(a, alpha, d, theta), ...] 을 차례로 곱한 T."""
    T = [[1.0 if i == j else 0.0 for j in range(4)] for i in range(4)]
    for row in rows:
        T = mul(T, dh(*row))
    return T


def yaw_of(T):
    """회전 블록 첫 열(끝 링크 x축)에서 자세각 atan2(T21, T11)."""
    return math.degrees(math.atan2(T[1][0], T[0][0]))


# ---------- 분리축 정리(SAT) ----------

def obb_corners(box):
    """box = (cx, cy, hw, hh, deg). 로컬(±hw, ±hh)를 회전 후 중심에 더한다."""
    cx, cy, hw, hh, deg = box
    c, s = math.cos(math.radians(deg)), math.sin(math.radians(deg))
    return [(cx + sx * c - sy * s, cy + sx * s + sy * c)
            for sx, sy in ((-hw, -hh), (hw, -hh), (hw, hh), (-hw, hh))]


def obb_axes(box):
    c, s = math.cos(math.radians(box[4])), math.sin(math.radians(box[4]))
    return [(c, s), (-s, c)]


def project(corners, axis):
    vals = [x * axis[0] + y * axis[1] for x, y in corners]
    return min(vals), max(vals)


def sat_obb(box1, box2):
    """네 축(두 상자의 로컬 x·y) 정사영 구간 비교. 반환: (충돌 여부, 축별 (구간1, 구간2, 겹침) 목록)."""
    c1, c2 = obb_corners(box1), obb_corners(box2)
    report = []
    hit = True
    for axis in obb_axes(box1) + obb_axes(box2):
        i1, i2 = project(c1, axis), project(c2, axis)
        overlap = not (i1[1] < i2[0] or i2[1] < i1[0])
        report.append((i1, i2, overlap))
        hit = hit and overlap
    return hit, report

File: ep09-transform-fk-wrapup/cpp/tf.cpp
// EP.9 지문 A~E + 경계 2문제 — C++ 최소 구현 (표준 라이브러리만).
// g++ -std=c++11 cpp/tf.cpp -o tf-cpp.exe; .\tf-cpp.exe   → python/run_all.py 와 같은 출력
#include <cmath>
#include <cstdio>
#include <vector>

namespace tf {
const double PI = 3.14159265358979323846;
double rad(double d) { return d * PI / 180.0; }
double deg(double r) { return r * 180.0 / PI; }

struct P2 { double x, y; };
struct P3 { double x, y, z; };
struct T4 { double m[4][4]; };

// ---------- 각도 도구 ----------
double wrap(double d) { double r = std::fmod(d + 180.0, 360.0); if (r < 0) r += 360.0; return r - 180.0; }
double circular_mean(const std::vector<double>& h) {
    double s = 0, c = 0;
    for (double a : h) { s += std::sin(rad(a)); c += std::cos(rad(a)); }
    return deg(std::atan2(s, c));
}

// ---------- 2D 강체변환 ----------
std::vector<P2> rigid2d(const std::vector<P2>& pts, double a, P2 t) {
    double c = std::cos(rad(a)), s = std::sin(rad(a));
    std::vector<P2> out;
    for (const P2& p : pts) out.push_back({c * p.x - s * p.y + t.x, s * p.x + c * p.y + t.y});
    return out;
}
void inv2d(double a, P2 t, double& ia, P2& it) {
    double c = std::cos(rad(a)), s = std::sin(rad(a));
    ia = -a; it = {-(c * t.x + s * t.y), -(-s * t.x + c * t.y)};
}

// ---------- 4x4 동차변환 ----------
T4 homog_z(double a, P3 t) { double c = std::cos(rad(a)), s = std::sin(rad(a)); return {{{c, -s, 0, t.x}, {s, c, 0, t.y}, {0, 0, 1, t.z}, {0, 0, 0, 1}}}; }
T4 homog_x(double a, P3 t) { double c = std::cos(rad(a)), s = std::sin(rad(a)); return {{{1, 0, 0, t.x}, {0, c, -s, t.y}, {0, s, c, t.z}, {0, 0, 0, 1}}}; }
T4 mul(const T4& a, const T4& b) {
    T4 o{};
    for (int i = 0; i < 4; ++i) for (int j = 0; j < 4; ++j) for (int k = 0; k < 4; ++k) o.m[i][j] += a.m[i][k] * b.m[k][j];
    return o;
}
P3 apply(const T4& t, P3 p) {
    double o[3];
    for (int i = 0; i < 3; ++i) o[i] = t.m[i][0] * p.x + t.m[i][1] * p.y + t.m[i][2] * p.z + t.m[i][3];
    return {o[0], o[1], o[2]};
}
T4 inv4(const T4& t) {
    T4 o{};
    for (int i = 0; i < 3; ++i) for (int j = 0; j < 3; ++j) o.m[i][j] = t.m[j][i];
    for (int i = 0; i < 3; ++i) o.m[i][3] = -(o.m[i][0] * t.m[0][3] + o.m[i][1] * t.m[1][3] + o.m[i][2] * t.m[2][3]);
    o.m[3][3] = 1.0;
    return o;
}

// ---------- 다링크 순기구학 ----------
std::vector<P2> fk(const std::vector<double>& links, const std::vector<double>& angles) {
    double x = 0, y = 0, cum = 0;
    std::vector<P2> pts{{0.0, 0.0}};
    for (size_t i = 0; i < links.size(); ++i) {
        cum += angles[i];
        x += links[i] * std::cos(rad(cum)); y += links[i] * std::sin(rad(cum));
        pts.push_back({x, y});
    }
    return pts;
}

// ---------- DH ----------
T4 dh(double a, double alpha, double d, double theta) {
    double ca = std::cos(rad(alpha)), sa = std::sin(rad(alpha)), ct = std::cos(rad(theta)), st = std::sin(rad(theta));
    return {{{ct, -st * ca, st * sa, a * ct}, {st, ct * ca, -ct * sa, a * st}, {0, sa, ca, d}, {0, 0, 0, 1}}};
}

// ---------- SAT ----------
struct OBB { double cx, cy, hw, hh, deg; };
std::vector<P2> obb_corners(const OBB& b) {
    const double sx[4] = {-1, 1, 1, -1}, sy[4] = {-1, -1, 1, 1};
    double c = std::cos(rad(b.deg)), s = std::sin(rad(b.deg));
    std::vector<P2> out;
    for (int i = 0; i < 4; ++i) { double lx = sx[i] * b.hw, ly = sy[i] * b.hh; out.push_back({b.cx + lx * c - ly * s, b.cy + lx * s + ly * c}); }
    return out;
}
void project(const std::vector<P2>& c, P2 ax, double& lo, double& hi) {
    lo = 1e300; hi = -1e300;
    for (const P2& p : c) { double v = p.x * ax.x + p.y * ax.y; if (v < lo) lo = v; if (v > hi) hi = v; }
}
}  // namespace tf

// -0.0000 방지용 출력 도우미
static void pf(double x) { if (std::fabs(x) < 5e-5) x = 0.0; std::printf("%.4f", x); }
static void p2(tf::P2 p) { std::printf("("); pf(p.x); std::printf(","); pf(p.y); std::printf(")"); }
static void p3(tf::P3 p) { std::printf("("); pf(p.x); std::printf(","); pf(p.y); std::printf(","); pf(p.z); std::printf(")"); }

static bool sat_obb(const tf::OBB& a, const tf::OBB& b) {
    using namespace tf;
    std::vector<P2> ca = obb_corners(a), cb = obb_corners(b);
    P2 axes[4] = {{std::cos(rad(a.deg)), std::sin(rad(a.deg))}, {-std::sin(rad(a.deg)), std::cos(rad(a.deg))},
                  {std::cos(rad(b.deg)), std::sin(rad(b.deg))}, {-std::sin(rad(b.deg)), std::cos(rad(b.deg))}};
    bool hit = true;
    for (int k = 0; k < 4; ++k) {
        double l1, h1, l2, h2;
        project(ca, axes[k], l1, h1); project(cb, axes[k], l2, h2);
        bool ov = !(h1 < l2 || h2 < l1);
        std::printf("  axis%d A=[", k + 1); pf(l1); std::printf(","); pf(h1); std::printf("] B=["); pf(l2); std::printf(","); pf(h2);
        std::printf("] %s\n", ov ? "overlap" : "separated");
        hit = hit && ov;
    }
    return hit;
}

int main() {
    using namespace tf;
    // ---------- A ----------
    std::printf("[A] #226 rigid2d rot=35deg t=(0.6,-0.4)\n");
    std::vector<P2> pts{{0.0, 0.0}, {1.2, 0.0}, {1.2, 0.8}, {0.0, 0.8}};
    double ang = 35.0; P2 t{0.6, -0.4};
    std::vector<P2> out = rigid2d(pts, ang, t);
    for (int i = 0; i < 4; ++i) { std::printf("P%d ", i); p2(pts[i]); std::printf(" -> "); p2(out[i]); std::printf("\n"); }
    double c = std::cos(rad(ang)), s = std::sin(rad(ang));
    std::printf("cos35="); pf(c); std::printf(" sin35="); pf(s); std::printf(" det="); pf(c * c - (-s) * s); std::printf(" area=0.9600\n");
    std::vector<P2> shifted;
    for (const P2& p : pts) shifted.push_back({p.x + t.x, p.y + t.y});
    std::printf("translate-then-rotate (wrong):");
    for (const P2& p : rigid2d(shifted, ang, {0.0, 0.0})) { std::printf(" "); p2(p); }
    std::printf("\n");
    double ia; P2 it; inv2d(ang, t, ia, it);
    std::printf("inv2d: rot="); pf(ia); std::printf("deg t'="); p2(it); std::printf("\n");
    std::printf("inv2d applied:");
    for (const P2& p : rigid2d(out, ia, it)) { std::printf(" "); p2(p); }
    std::printf("\n");
    double vx = out[1].x - t.x, vy = out[1].y - t.y;
    double by_atan2 = deg(std::atan2(vy, vx)) - deg(std::atan2(pts[1].y, pts[1].x));
    double by_acos = deg(std::acos((pts[1].x * vx + pts[1].y * vy) / (std::hypot(pts[1].x, pts[1].y) * std::hypot(vx, vy))));
    std::printf("recover_angle from P1'="); p2(out[1]); std::printf(": atan2="); pf(by_atan2); std::printf("deg acos="); pf(by_acos);
    std::printf("deg wrap(-325)="); pf(wrap(-325.0)); std::printf("deg\n");

    // ---------- B ----------
    std::printf("[B] #258 chain T_AC = T_AB * T_BC\n");
    T4 T_AB = homog_z(60.0, {0.5, 0.2, 0.0}), T_BC = homog_x(90.0, {0.0, 0.4, 0.3});
    P3 p_C{0.2, 0.1, 0.5};
    T4 T_AC = mul(T_AB, T_BC);
    std::printf("t_AC = "); p3({T_AC.m[0][3], T_AC.m[1][3], T_AC.m[2][3]}); std::printf("\n");
    std::printf("naive sum (wrong) = "); p3({0.5 + 0.0, 0.2 + 0.4, 0.0 + 0.3}); std::printf("\n");
    P3 p_B = apply(T_BC, p_C); std::printf("p_B = "); p3(p_B); std::printf("\n");
    P3 p_A = apply(T_AC, p_C); std::printf("p_A via T_AC = "); p3(p_A); std::printf("\n");
    std::printf("p_A via two steps = "); p3(apply(T_AB, p_B)); std::printf("\n");
    std::printf("reversed order T_BC*T_AB (wrong) = "); p3(apply(mul(T_BC, T_AB), p_C)); std::printf("\n");
    std::printf("inv4(T_AC) p_A = "); p3(apply(inv4(T_AC), p_A)); std::printf("\n");

    // ---------- C ----------
    std::printf("[C] #221 fk links=(1.4,0.9) angles=(25,50)\n");
    std::vector<P2> j = fk({1.4, 0.9}, {25.0, 50.0});
    for (size_t i = 0; i < j.size(); ++i) { std::printf("j%d ", (int)i); p2(j[i]); std::printf("\n"); }
    std::printf("cumulative = 25.0000/75.0000 reach = "); pf(std::hypot(j[2].x, j[2].y)); std::printf("\n");
    std::printf("absolute-angle mistake (wrong) = "); p2({j[1].x + 0.9 * std::cos(rad(50.0)), j[1].y + 0.9 * std::sin(rad(50.0))}); std::printf("\n");
    std::printf("[C] #220 fk links=(1.2,0.9,0.6) angles=(20,35,-25)\n");
    std::vector<P2> j3 = fk({1.2, 0.9, 0.6}, {20.0, 35.0, -25.0});
    for (size_t i = 0; i < j3.size(); ++i) { std::printf("j%d ", (int)i); p2(j3[i]); std::printf("\n"); }
    std::printf("cumulative = 20.0000/55.0000/30.0000\n");

    // ---------- D ----------
    std::printf("[D] #237 dh chain a=(1.4,0.9) theta=(25,50)\n");
    T4 A1 = dh(1.4, 0.0, 0.0, 25.0), A2 = dh(0.9, 0.0, 0.0, 50.0);
    std::printf("A1 R=["); pf(A1.m[0][0]); std::printf(","); pf(A1.m[0][1]); std::printf(";"); pf(A1.m[1][0]); std::printf(","); pf(A1.m[1][1]); std::printf("] t="); p2({A1.m[0][3], A1.m[1][3]}); std::printf("\n");
    std::printf("A2 R=["); pf(A2.m[0][0]); std::printf(","); pf(A2.m[0][1]); std::printf(";"); pf(A2.m[1][0]); std::printf(","); pf(A2.m[1][1]); std::printf("] t="); p2({A2.m[0][3], A2.m[1][3]}); std::printf("\n");
    T4 T = mul(A1, A2);
    std::printf("T=A1*A2 -> "); p2({T.m[0][3], T.m[1][3]}); std::printf(" yaw=atan2(T21,T11)="); pf(deg(std::atan2(T.m[1][0], T.m[0][0]))); std::printf("deg\n");
    T4 T237 = mul(dh(2.0, 0.0, 0.0, deg(0.5236)), dh(1.5, 0.0, 0.0, deg(0.7854)));
    std::printf("original #237 a=(2.0,1.5) theta=(0.5236,0.7854)rad -> "); p2({T237.m[0][3], T237.m[1][3]}); std::printf("\n");

    // ---------- E ----------
    std::printf("[E] #223 sweep theta1=25 theta2=0..90 step 15\n");
    int n = 0;
    for (int a = 0; a <= 90; a += 15) {
        P2 e = fk({1.4, 0.9}, {25.0, (double)a})[2];
        std::printf("%02d ", a); p2(e); std::printf(" r_from_j1="); pf(std::hypot(e.x - j[1].x, e.y - j[1].y));
        std::printf(" dist="); pf(std::hypot(e.x, e.y)); std::printf("\n");
        ++n;
    }
    std::printf("points = %d center j1 = ", n); p2(j[1]); std::printf(" annulus inner=0.5000 outer=2.3000\n");

    // ---------- #512 ----------
    std::printf("[#512] SAT obb A=(0,0,0.6,0.4,0deg) B=(1.1,0.35,0.5,0.3,30deg)\n");
    OBB A{0.0, 0.0, 0.6, 0.4, 0.0};
    for (double cx : {1.1, 1.45}) {
        OBB B{cx, 0.35, 0.5, 0.3, 30.0};
        std::printf("B cx="); pf(cx); std::printf(" corners:");
        for (const P2& p : obb_corners(B)) { std::printf(" "); p2(p); }
        std::printf("\n");
        std::printf("  collide = %s\n", sat_obb(A, B) ? "True" : "False");
    }

    // ---------- #558 ----------
    std::printf("[#558] circular mean headings=(10,350,20)\n");
    std::vector<double> h{10.0, 350.0, 20.0};
    std::printf("simple mean (wrong) = "); pf((h[0] + h[1] + h[2]) / 3.0); std::printf("\n");
    std::printf("circular mean = "); pf(circular_mean(h)); std::printf("deg\n");
    return 0;
}
